pkg/stats/collector.go

```diff
@@ -142,49 +158,88 @@ func (s *BufferedCollector) incr(k primaryKey) {
 }
 
 func (s *BufferedCollector) send(metrics []Metric) {
-	if len(metrics) == 0 {
-		return
-	}
-	ctx, cancel := context.WithTimeout(context.Background(), s.sendTimeout)
-	defer cancel()
-	err := s.sender.SendEvent(ctx, s.getInstallationID(), s.processID, metrics)
-	if err != nil {
-		logging.Default().
-			WithError(err).
-			WithField("service", "stats_collector").
-			Debug("could not send stats")
-	}
+	s.pendingRequests.Add(1)
+	go func() {
+		defer s.pendingRequests.Done()
+		if len(metrics) == 0 {
+			return
+		}
+		ctx, cancel := context.WithTimeout(context.Background(), s.sendTimeout)
+		defer cancel()
+		err := s.sender.SendEvent(ctx, s.getInstallationID(), s.processID, metrics)
+		if err != nil {
+			logging.Default().
+				WithError(err).
+				WithField("service", "stats_collector").
+				Debug("could not send stats")
+		}
+	}()
 }
 
 func (s *BufferedCollector) CollectEvent(class, action string) {
+	if s.isCtxCancelled() {
+		return
+	}
+	s.pendingWrites.Add(1)
+	defer s.pendingWrites.Done()
+
 	s.writes <- primaryKey{
 		class:  class,
 		action: action,
 	}
 }
 
-func (s *BufferedCollector) Done() <-chan bool {
-	return s.done
+func (s *BufferedCollector) isCtxCancelled() bool {
+	return atomic.LoadInt32(&s.ctxCancelled) == 1
 }
 
 func (s *BufferedCollector) Run(ctx context.Context) {
-	go s.collectHeartbeat(ctx)
-	for {
-		select {
-		case w := <-s.writes: // collect events
-			s.incr(w)
-		case <-s.flushTicker.Tick(): // every N seconds, send the collected events
-			s.handleRuntimeStats()
-			metrics := makeMetrics(s.cache)
-			s.cache = make(keyIndex)
-			go s.send(metrics) // no need to block on this
-		case <-ctx.Done(): // we're done
-			metrics := makeMetrics(s.cache)
-			s.send(metrics)
-			s.done <- true
-			return
+	atomic.StoreInt32(&s.runCalled, 1)
+	go func() {
+		for {
+			select {
+			case w := <-s.writes: // collect events
+				s.incr(w)
+			case <-s.heartbeatTicker.Tick():
+				s.CollectEvent("global", "heartbeat")
```

Comment:
Use `s.incr` to update the cache with heartbeat information - at this scope this the recommended place to update the cache.
Collect event writes to the channel this goroutine processes. It can be tricky if the channel is full, and we try to send something while we can't process - it will block the code forever.